src/DIRC_geo.cpp

```diff
@@ -108,46 +108,42 @@ static Ref_t createDetector(Detector& desc, xml_h e, SensitiveDetector sens)
-  double ztrans2 = lens_thickness / 2. + sqrt(lens_r2 * lens_r2 - lens_height / 2. * lens_height / 2.) - layer12;
+  double thight = bar_height;
+  double cr2 = sqrt(lens_width * lens_width / 4. + thight * thight / 4.);
 
-  Box gfbox("fbox", 0.5 * prism_short_edge, 0.5 * prism_width, 0.5 * lens_thickness);
-  Box gcbox("cbox", 0.5 * prism_short_edge, 0.5 * prism_width + 1 * mm, 0.5 * lens_thickness);
+  double lens_min_thickness = 2.0 * mm;
 
-  Position         tTrans1(0.5 * (prism_short_edge + lens_height), 0, lens_thickness - layer12);
-  Position         tTrans0(-0.5 * (prism_short_edge + lens_height), 0, lens_thickness - layer12);
-  SubtractionSolid tubox("tubox", gfbox, gcbox, tTrans1);
-  SubtractionSolid gubox("gubox", tubox, gcbox, tTrans0);
+  double ztrans1 = -lens_thickness / 2. - sqrt(lens_r1 * lens_r1 - cr2 * cr2) + lens_min_thickness;
+  double ztrans2 = -lens_thickness / 2. - sqrt(lens_r2 * lens_r2 - cr2 * cr2) + lens_min_thickness * 2;
 
-  Tube gcylinder1("Cylinder1", 0, lens_r1, 0.5 * prism_width, 0 * deg, 360 * deg);
-  Tube gcylinder2("Cylinder2", 0, lens_r2, 0.5 * prism_width - 0.5 * mm, 0 * deg, 360 * deg);
-  Tube gcylinder1c("Cylinder1c", 0, lens_r1, 0.5 * prism_width + 0.5 * mm, 0 * deg, 360 * deg);
-  Tube gcylinder2c("Cylinder2c", 0, lens_r2, 0.5 * prism_width + 0.5 * mm, 0 * deg, 360 * deg);
+  Box gfbox("Fbox", 0.5 * prism_short_edge, 0.5 * lens_width, 0.5 * lens_thickness);
+  Tube gfstube("ftube", 0, cr2, 0.5 * lens_thickness, 0 * deg, 360 * deg);
 
-  IntersectionSolid lens_layer1_solid("lens_layer1_solid", gubox, gcylinder1,
-                                      Transform3D(RotationX(M_PI / 2.), Position(0, 0, ztrans1)));
-  SubtractionSolid  gLenst("temp", gubox, gcylinder1c, Transform3D(RotationX(M_PI / 2.), Position(0, 0, ztrans1)));
+  Sphere gsphere1("Sphere1", 0, lens_r1, 0 * deg, 180 * deg, 0 * deg, 360 * deg);
+  Sphere gsphere2("Sphere2", 0, lens_r2, 0 * deg, 180 * deg, 0 * deg, 360 * deg);
 
-  IntersectionSolid lens_layer2_solid("lens_layer2_solid", gLenst, gcylinder2,
-                                      Transform3D(RotationX(M_PI / 2.), Position(0, 0, ztrans2)));
-  SubtractionSolid  lens_layer3_solid("lens_layer3_solid", gLenst, gcylinder2c,
-                                      Transform3D(RotationX(M_PI / 2.), Position(0, 0, ztrans2)));
+  IntersectionSolid gbbox("bbox", gfbox, gfbox, Position(0, 0, -lens_min_thickness * 2));
+  IntersectionSolid gsbox("sbox", gfstube, gfbox, Position(0, 0, lens_min_thickness * 2));
+
+  UnionSolid gubox("unionbox", gbbox, gsbox);
+
+  IntersectionSolid lens_layer1_solid("lens_layer1_solid", gubox, gsphere1, Position(0, 0, -ztrans1));
+  SubtractionSolid  gLenst("temp", gubox, gsphere1, Position(0, 0, -ztrans1));
+
+  IntersectionSolid lens_layer2_solid("lens_layer2_solid", gLenst, gsphere2, Position(0, 0, -ztrans2));
+  SubtractionSolid  lens_layer3_solid("lens_layer3_solid", gLenst, gsphere2, Position(0, 0, -ztrans2));
```

Comment:
This is using way too many boolean solids of boolean solids. There should be one volume that is the volume of the 3 layers combined (not a union: just a box or a tube or whatever it is here), with material Quartz. Then inside that volume you should place the middle lens layer: a subtraction of two spheres, then an intersection with the outer volume.